### PR TITLE
Consider generics when resolving Serdes for KafkaStreams binder

### LINKED ISSUE #2318: Kafka Streams Serdes having nested generic doesn't work
@AnandMandhre commented on [Sun Mar 06 2022](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/issues/1205)

Two functions on kafka topics with nested generics like below causes incorrect serde to be selected during deserialization.
```
    @Bean
    public Function<KStream<String, CloudEvent<ClassA>>, KStream<String, CloudEvent<ClassB>>> method1() {
         ....... //lambda
    }

    @Bean
    public Function<KStream<String, CloudEvent<ClassB>>, KStream<String, CloudEvent<ClassC>>> method2() {
        ...... //lambda
    }

   @Bean
    public Serde<CloudEvent<ClassA>> classASerde(ObjectMapper mapper, Validator validator) {     
        return StreamsSerdes.classASerde(mapper,validator);
    }

    @Bean
    public Serde<CloudEvent<ClassB>> classBSerde(ObjectMapper mapper, Validator validator) {     
        return StreamsSerdes.classBSerde(mapper,validator);
    }

```
See also https://stackoverflow.com/questions/71351893/kafka-streams-serdes-having-nested-generic-doesnt-work

### PR DESCRIPTION
**Pull request by onobc**
This proposal fixes Serde bean resolution to include parameterized types when matching. 

#### Highlights
- It also includes a new utility class that is also capable of finding a list of matching serdes for a target type, ordered by specificity. 

Fixes #2318

### CODE REVIEW

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/KeyValueSerdeResolver.java [hunk 1]
```diff
@@ -286,101 +276,6 @@ private boolean isResolvalbeKafkaStreamsType(ResolvableType resolvableType) {
 				GlobalKTable.class.isAssignableFrom(resolvableType.getRawClass()));
 	}
 
-	private Serde<?> getSerde(ResolvableType generic, Serde<?> fallbackSerde) {
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/KeyValueSerdeResolver.java:289]: Pulled this out into SerdeResolverUtils

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+			if (!genericRawClazz.isAssignableFrom((Object.class))) {
+				serde = new JsonSerde(genericRawClazz);
+			}
+		}
+		return serde;
+	}
+
+	private static boolean isSerdeFromStandardDefaults(Serde<?> serde) {
+		if (serde != null) {
+			if (Number.class.isAssignableFrom(serde.getClass())) {
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:113]: **[UNRELATED]** I noticed this does not really do anything.

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+			// If the type is Object, then skip assigning the JsonSerde and let the fallback mechanism takes precedence.
+			if (!genericRawClazz.isAssignableFrom((Object.class))) {
+				serde = new JsonSerde(genericRawClazz);
+			}
+		}
+		return serde;
+	}
+
+	private static boolean isSerdeFromStandardDefaults(Serde<?> serde) {
+		if (serde != null) {
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:112]: **[UNRELATED]** I am going to do a follow-on PR to rework this default fallback portion of the code
- invert the if (short circuit) 
- slightly change how the "is default serdes" is implemented

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtilsTests.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	 *                    | GE<Date> | GE<? extends Date> | GE<?> | GE
+	 * ------------------------------------------------------------------
+	 * GE<Date>           | Y        | Y                  | Y     | Y
+	 * GE<? extends Date> | N        | Y                  | Y     | N
+	 * GE<?>              | N        | N                  | Y     | N
+	 * GE                 | N        | N                  | N     | N
+	 * ------------------------------------------------------------------
+	 * }</pre>
+	 */
+	@Test
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtilsTests.java:59]: In follow-on PR I will also beef up this test to include a more complicated Serde type that has 2 level generics as well as some edge-casey like checks.

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/function/SerdesProvidedAsBeansTests.java [hunk 5]
```diff
@@ -42,78 +43,178 @@
...
+ */
+@EmbeddedKafka(topics = { "topic1", "topic2" })
 public class SerdesProvidedAsBeansTests {
 
 	private static final EmbeddedKafkaBroker embeddedKafka = EmbeddedKafkaCondition.getBroker();
 
 	@Test
-	void testKstreamWordCountFunction() throws NoSuchMethodException {
-		SpringApplication app = new SpringApplication(SerdeProvidedAsBeanApp.class);
+	void simpleSerdeBeansAreResolvedProperly() throws Exception {
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/function/SerdesProvidedAsBeansTests.java:61]: I ended up adding to this related existing test rather than have 1 for generics and 1 for simple types. There is some duplicated code I will refactor when I do that follow on PR I mention in the other comments.

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+	 *  {@code fallbackSerde}, or finally a standard default serde if no fallback specified.
+	 *
+	 * @param context the application context
+	 * @param targetType the target type to find the serde for
+	 * @param fallbackSerde the fallback serde in case no matching serde bean found in the context
+	 * @return serde to use for the target type
+	 */
+	static Serde<?> resolveForType(ConfigurableApplicationContext context, ResolvableType targetType, Serde<?> fallbackSerde) {
+
+		List<Serde<?>> matchingSerdes = findMatchingSerdes(context, targetType);
```
Comment (onobc) [inline binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:62]: I apologize in advance for the drastic split/refactor on a single commit. This class is largely the previous method in KeyValueResolver w/ the added "findMatching" API. It is largely unchanged in this method below this point.

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java (hunk 3 above)
Comment (onobc) [reply binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:112]: To be handled in #2370

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtilsTests.java (hunk 4 above)
Comment (onobc) [reply binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtilsTests.java:59]: To be handled in #2370

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/function/SerdesProvidedAsBeansTests.java (hunk 5 above)
Comment (onobc) [reply binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/test/java/org/springframework/cloud/stream/binder/kafka/streams/function/SerdesProvidedAsBeansTests.java:61]: To be handled in #2370

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		if (!matchingSerdes.isEmpty()) {
+			return matchingSerdes.get(0);
+		}
+
+		// We don't attempt to find a matching Serde for type '?'
+		if (targetType.getRawClass() == null) {
+			return null;
```
[the remark this answers is not in the document]
Comment (onobc) [reply binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:69]: To be handled in #2370

File: binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	 * @return list of matching serdes order by most specific match, or an empty list if no matches found
+	 */
+	static List<Serde<?>> findMatchingSerdes(ConfigurableApplicationContext context, ResolvableType targetType) {
+		// We don't attempt to find a matching Serde for type '?'
+		if (targetType.getRawClass() == null) {
+			return Collections.emptyList();
+		}
+		List<SerdeWithSpecificityScore> matchingSerdes = new ArrayList<>();
+
+		context.getBeansOfType(Serde.class).forEach((beanName, serdeBean) -> {
```
[the remark this answers is not in the document]
Comment (onobc) [reply binders/kafka-binder/spring-cloud-stream-binder-kafka-streams/src/main/java/org/springframework/cloud/stream/binder/kafka/streams/SerdeResolverUtils.java:143]: To be handled in #2370

**Comment by sobychacko** [comment]
Merged upstream to main via https://github.com/spring-cloud/spring-cloud-stream/commit/210eb7e9f56ae476ab26fcae015f8e3a29738a7a.
